sliver/handlers/generic-tun-handlers.go

```diff
@@ -177,3 +189,92 @@ func shellReqHandler(envelope *sliverpb.Envelope, connection *transports.Connect
 	// {{end}}
 
 }
+
+func tcpTunnelReqHandler(envelope *sliverpb.Envelope, connection *transports.Connection) {
+	fmt.Println("A call to tcpTunnelReqHandler()")
+
+	returnStatusCode := func(statusCode byte, connection *transports.Connection) {
+		// {{if .Debug}}
+		log.Printf("Returning status code for tcptunnel %d\n", statusCode)
+		log.Printf("Connection object %s\n", connection)
+		// {{end}}
+		tcpTunnelResp, _ := proto.Marshal(&sliverpb.TCPTunnel{
+			StatusCode: uint32(statusCode),
+		})
+		connection.Send <- &sliverpb.Envelope{
+			ID: envelope.ID,
+			Data: tcpTunnelResp,
+		}
+	}
+
+	tcpTunnelReq := &sliverpb.TCPTunnelReq{}
+	err := proto.Unmarshal(envelope.Data, tcpTunnelReq)
+	if err != nil {
+		returnStatusCode(failedMessageUnmarshalErrorCode, connection)
+		return
+	}
+
+	remoteHost := tcpTunnelReq.RemoteHost
+	remotePort := tcpTunnelReq.RemotePort
+
+	remoteAddressString := fmt.Sprintf("%s:%d", remoteHost, remotePort)
+
+	remoteAddress, err := net.ResolveTCPAddr("tcp4", remoteAddressString)
+	if err != nil {
+		returnStatusCode(failedTCPRemoteHostResolve, connection)
+		return
+	}
+
+	remoteConn, err := net.DialTCP("tcp4", nil, remoteAddress)
+	if err != nil {
+		returnStatusCode(failedTCPRemoteHostConnect, connection)
+		return
+	}
+	returnStatusCode(successfulTCPTunnelReq, connection)
+
+	tunnel := &transports.Tunnel{
+		ID:     tcpTunnelReq.TunnelID,
+		Reader: remoteConn,
+		Writer: remoteConn,
+	}
+	connection.AddTunnel(tunnel)
+
+	go func() {
+		for connection.Tunnel(tunnel.ID) != nil {
+			tWriter := tunnelWriter{
+				tun:  tunnel,
+				conn: connection,
+			}
+			//bytesRead, err := io.Copy(tWriter, tunnel.Reader)
+			byteArrayRead := make([]byte, 1024)
+			bytesRead, err := tunnel.Reader.Read(byteArrayRead)
+			if bytesRead != 0 {
+				fmt.Printf("Read %d bytes from socket\n", bytesRead)
+				tWriter.Write(byteArrayRead[:bytesRead])
+			} else if err != nil && strings.Contains(err.Error(), "An existing connection was forcibly closed by the remote host") {
+				// Socket has been closed by remote host
+				// TODO : Find a better way to do this
+				break
+			}
+
+			if err != nil && err != io.ErrShortWrite {
+				fmt.Printf("Closing tunnel because of error %s\n", err.Error())
+				break
+			}
+		}
+
+		// Cleanup
```

Comment:
@rkervella Hi, I am having some trouble with closing my tunnel from the client, would you mind looking at it ?

In this code snippet the client should send a `MsgTunnelClose` , according to my understanding the function `tunnelCloseHandler` should handle messages of type `MsgTunnelClose` and therefore close the tunnel (which would lead to the remote socket being closed).
But for some reason this message is handled by `tunnelDataHandler` and the tunnel stays open 😢.
I tried searching for other usages of the rpc function `CloseTunnel` but with no success.

Reply:
Gonna have a look.

Reply:
Thanks 😄

Reply:
Your `rpc.CloseTunnel` call should indeed arrive in `tunnelCloseHandler`, I don't see any reason why it would not. However, I just noticed that `Tunnel.Close()` first sends a `MsgTunnelClose` message, and then some data:

https://github.com/BishopFox/sliver/blob/1e6abf53bb60c02ff8a6835569b35a29dd57069f/server/core/tunnels.go#L98

So my guess would be that what you interpret as `tunnelDataHandler` receives the `MsgTunnelClose` message is in fact a copy of this empty data that is sent in the above call.

And this channel is read in a loop here: https://github.com/BishopFox/sliver/blob/1e6abf53bb60c02ff8a6835569b35a29dd57069f/server/rpc/rpc-tunnel.go#L112-L136

@moloch-- might be able to help in this case.

Reply:
Ok, that explains it.
So can I replace `tunnel.ToImplant <- data` with this code snippet ?
```
session := Sessions.Get(tunnel.SessionID)
if session == nil {
	return ErrInvalidSessionID
}
session.Send <- &sliverpb.Envelope{
	Type: sliverpb.MsgTunnelClose,
	Data: tunnelClose,
}
```

I tested this and it fixes my bug, is there anything that it might break ?

Reply:
The only feature that would break would be `shell`. So if that still works, I'd say we should be good.

Reply:
It's seems to work fine so I guess it's okay

Reply:
Great :)